crates/rover-client/src/operations/persisted_queries/publish/types.rs

```diff
@@ -107,6 +113,145 @@ impl From<PersistedQueriesPublishInput> for QueryVariables {
+            let ast = Parser::new(&body).parse();
+
+            let operation_definitions: Vec<OperationDefinition> = ast
+                .clone()
+                .document()
+                .definitions()
+                .filter_map(|definition| {
+                    if let Definition::OperationDefinition(operation_definition) = definition {
+                        Some(operation_definition)
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+
+            let maybe_definition = match operation_definitions.len() {
+                0 => {
+                    ids_with_no_operations.push(id.clone());
+                    None
+                }
+                1 => {
+                    if let Some(operation_definition) = operation_definitions.get(0) {
+                        Some(operation_definition)
+                    } else {
+                        ids_with_no_operations.push(id.clone());
+                        None
+                    }
+                }
+                _ => {
+                    ids_with_multiple_operations.push(id.clone());
+                    None
+                }
+            };
+
+            if let Some(operation_definition) = maybe_definition {
+                // attempt to extract operation type, defaulting to "query" if we can't find one
+                let operation_type = match operation_definition.operation_type() {
+                    Some(operation_type) => {
+                        match (
+                            operation_type.mutation_token(),
+                            operation_type.query_token(),
+                            operation_type.subscription_token(),
+                        ) {
+                            (Some(_mutation), _, _) => PersistedQueryOperationType::Mutation,
+                            (_, Some(_query), _) => PersistedQueryOperationType::Query,
+                            (_, _, Some(_subscription)) => {
+                                PersistedQueryOperationType::Subscription
+                            }
+                            // this should probably be unreachable, but just default to query regardless
+                            _ => PersistedQueryOperationType::Query,
+                        }
+                    }
+                    None => PersistedQueryOperationType::Query,
+                };
+
+                // track valid operations and the IDs of invalid operations
+                if let Some(operation_name) = operation_definition.name() {
+                    operations.push(PersistedQueryOperation {
+                        name: operation_name.text().to_string(),
+                        r#type: operation_type,
+                        body: body.to_string(),
+                        id: id.to_string(),
+                    });
+                } else {
+                    let mut parse_errors = ast.errors().peekable();
```

Comment:
I still think it's a surprising choice to only look for syntax errors at the end here instead of every time; it does seem  at least worthy of a comment?